meshdb/meshdb.go

```diff
@@ -318,26 +328,93 @@ func (m *MeshDB) FindMiniHeaderByBlockNumber(blockNumber *big.Int) (*miniheader.
 	return miniHeaders[0], nil
 }
 
+// UpdateMiniHeaderRetentionLimit updates the MiniHeaderRetentionLimit. This is only used by tests in order
+// to set the retention limit to a smaller size, making the tests shorter in length
+func (m *MeshDB) UpdateMiniHeaderRetentionLimit(limit int) error {
+	m.MiniHeaderRetentionLimit = limit
+	return m.PruneMiniHeadersAboveRetentionLimit()
+}
+
+// PruneMiniHeadersAboveRetentionLimit prunes miniHeaders from the DB that are above the retention limit
+func (m *MeshDB) PruneMiniHeadersAboveRetentionLimit() error {
+	if totalMiniHeaders, err := m.MiniHeaders.Count(); err != nil {
+		return err
+	} else if totalMiniHeaders > m.MiniHeaderRetentionLimit {
+		miniHeadersToRemove := totalMiniHeaders - m.MiniHeaderRetentionLimit
+		log.WithFields(log.Fields{
+			"numHeadersToRemove": miniHeadersToRemove,
+			"totalHeadersStored": totalMiniHeaders,
+		}).Warn("Removing outdated block headers in database (this can take a while)")
+		latestMiniHeader, err := m.FindLatestMiniHeader()
+		if err != nil {
+			return err
+		} else if latestMiniHeader != nil {
+			minBlockNumber := big.NewInt(0).Sub(latestMiniHeader.Number, big.NewInt(int64(m.MiniHeaderRetentionLimit)-1))
+			if err := m.ClearOldMiniHeaders(minBlockNumber); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
 // ClearAllMiniHeaders removes all stored MiniHeaders from the database.
 func (m *MeshDB) ClearAllMiniHeaders() error {
+	return m.clearMiniHeadersWithFilter(m.MiniHeaders.numberIndex.All())
+}
+
+// ClearOldMiniHeaders removes all stored MiniHeaders with a block number less then
+// the given minBlockNumber.
+func (m *MeshDB) ClearOldMiniHeaders(minBlockNumber *big.Int) error {
+	filter := m.MiniHeaders.numberIndex.RangeFilter(
+		uint256ToConstantLengthBytes(big.NewInt(0)),
+		uint256ToConstantLengthBytes(minBlockNumber),
+	)
+	return m.clearMiniHeadersWithFilter(filter)
+}
+
+func (m *MeshDB) clearMiniHeadersWithFilter(filter *db.Filter) error {
+	for page := 0; ; page++ {
+		removed, err := m.clearMiniHeadersForPage(filter, page)
+		if err != nil {
+			return err
+		}
+		if removed == 0 {
+			break
+		}
+	}
+	return nil
+}
+
+func (m *MeshDB) clearMiniHeadersForPage(filter *db.Filter, page int) (removed int, err error) {
 	txn := m.MiniHeaders.OpenTransaction()
 	defer func() {
 		_ = txn.Discard()
 	}()
-	var storedHeaders []*miniheader.MiniHeader
-	if err := m.MiniHeaders.FindAll(&storedHeaders); err != nil {
-		return err
-	}
-	for _, header := range storedHeaders {
-		if err := txn.Delete(header.ID()); err != nil {
-			return err
+	log.WithFields(log.Fields{
+		"page":    page,
+		"perPage": miniHeadersMaxPerPage,
+	}).Trace("Removing MiniHeaders")
+	offset := page * miniHeadersMaxPerPage
+	var miniHeaders []*miniheader.MiniHeader
+	if err := m.MiniHeaders.NewQuery(filter).Offset(offset).Max(miniHeadersMaxPerPage).Run(&miniHeaders); err != nil {
```

Comment:
We need to remove `Offset` unless I'm missing something. Since we're deleting miniheaders, the offset should always be `0`, or else we'll end up with almost 70000 miniheaders.

Reply:
@albrow can you confirm? I didn't actually write this code.

Reply:
Yes that's correct. I already mentioned it in Slack. This was a bug that I accidentally introduced in subsequent commits within this PR when I changed from using a single transaction to remove all the MiniHeaders to multiple transactions.

Reply:
@fabioberger can you also improve the test so that it would catch this bug? We need to add more than `miniHeadersMaxPerPage` in order to test that pagination works correctly and all the MiniHeaders are removed. To make testing this case easier/faster, you can unit test `ClearOldMiniHeaders` in parallel without setting up block watcher or order watcher.

Reply:
We can also refactor this a bit since we no longer need to pass in the `page` number. And then, we should probably rename `clearMiniHeadersForPage`.

Reply:
I added this test and applied the other recommended changes.